### ISSUE TITLE
dict_keys is identified as hashable

### ISSUE THREAD

**Issue by headtr1ck** [python/mypy#14674]
**Bug Report**

Mypy seems to think that dict_keys (`{}.keys()`) are hashable, which they are not.

**To Reproduce**

```python
def f(a: Hashable) -> int:
    return 1
f({}.keys())  # does not raise an issue
```

**Expected Behavior**

Since dict-keys are not hashable (`hash({}.keys())` fails and `isinstance({}.keys(), Hashable) == False`) I would expect mypy to throw an error in the above example.

**Actual Behavior**

> Success: no issues found in 1 source file

**Your Environment**

- Mypy version used: 0.982
- Mypy command-line flags: /
- Mypy configuration options from `mypy.ini` (and other config files): /
- Python version used: 3.10.8

**Comment by erictraut** [comment]
While it's true that `dict_key` objects are not hashable at runtime, the class does conform to the `Hashable` protocol — at least as it's defined in the typeshed stubs. That means mypy is arguably correct in not emitting an error for the above code. 

Hashability is difficult to enforce using the type system features because subclasses of a hashable class can be non-hashable.
